### PR TITLE
Add Docs (based on yard)

### PR DESCRIPTION
**Pull request by movitto**

### CODE REVIEW

**Comment by movitto** [comment]
@roliveri please review, just rdoc / yardoc additions, no functional changes. Would like to merge b4 it gets stale

**Comment by roliveri** [comment]
@movitto Sorry, I should have merged this before my other PR. You'll have to rebase, but it doesn't look like the conflicts are too bad.

File: lib/virtfs/context.rb [hunk 1]
```diff
@@ -1,4 +1,10 @@
 module VirtFS
+  # FS-specific settings under which FS calls occur.
```
Comment (roliveri) [inline lib/virtfs/context.rb]: Actually the context maintains FS-specific _state_. That _state_ includes: what filesystems are mounted and where, the current working directory, and the current root.

File: lib/virtfs/context.rb [hunk 2]
```diff
@@ -1,4 +1,10 @@
...
+  #
+  # VirtFS maps an independent context instance to each
+  # mounted FS instance, and internally switches to it before
+  # dispatching target FS calls. This class implements the
```
Comment (roliveri) [inline lib/virtfs/context.rb]: There is not an independent context per mounted FS instance. There is a context per thread group of the process - so normally, there will only be one context per process.

Due to the nature of constants in ruby, VirtFS _activation_ must be a process-while operation. When a given application requires a process to have multiple _views_ of the current FS state, multiple contexts can be used. By dividing the process into multiple thread-groups, each thread-group can have it's own view of the current FS state. For example, one group might only see the native filesystem, while others can see additional filesystems, or totally different filesystems. Basing contexts on thread-groups, as opposed to individual threads, permits multi-threaded filesystem access when necessary.

File: lib/virtfs/context.rb [hunk 3]
```diff
@@ -14,13 +20,28 @@ def initialize
...
       @dir_mutex.synchronize do
         raise "Context already assigned to key: #{@key}" if !@key.nil? && !val.nil?
         @key = val
       end
     end
 
+    # Mount the specified FS instance at the specified mount point.
+    # This registers specified fs to be accessed through the specified mount
+    # point via internal mechanisms. After this point any calls to this mount
+    # point through VirtFS will be mapped through the specified fs instance
```
Comment (roliveri) [inline lib/virtfs/context.rb]: Should be:
...through VirtFS, **under this context**, will be mapped...

File: lib/virtfs/context.rb [hunk 4]
```diff
@@ -127,13 +166,23 @@ def chdir(dir)
...
+    # @param raise_full_path [Boolean] indicates if error should be raised if lookup fails
+    # @param include_last [Boolean] indicates if last path component should be returned
+    #
+    # @raise [RunTimeError] if path could not be looked up and raise_full_path is true
+    # @raise [SystemCallError] if path could not be looked up
+    #
+    # @see #mount_lookup
+    # @see #expand_links
+    #
     def path_lookup(path, raise_full_path = false, include_last = true)
```
Comment (roliveri) [inline lib/virtfs/context.rb]: The methods defined from here to the end of file are not part of the public VirtFS interface (they are for internal use). I'm not sure if they should be included in the documentation - or minimally, it should be made clear they are for internal use only.

File: lib/virtfs/find_class_methods.rb [hunk 5]
```diff
@@ -1,11 +1,17 @@
...
+  # VirtFS Find Class representation - implements the core Ruby Find methods,
+  # dispatching to underlying mounted VirtFS filesystems
   module FindClassMethods
     #
     # Modified version of Find.find:
     # - Accepts only a single path.
     # - Can be restricted by depth - optimization for glob searches.
     # - Will work with VirtFS, even when it's not active.
     #
+    # @param path [String] path to lookup
```
Comment (roliveri) [inline lib/virtfs/find_class_methods.rb]: **path** is the starting directory of the _find_.

File: lib/virtfs/find_class_methods.rb [hunk 6]
```diff
@@ -50,6 +59,11 @@ def glob_str?(str)
       str.gsub(/\\./, "X").count(GLOB_CHARS) != 0
     end
 
+    # Returns files matching glob pattern
+    #
+    # @param glob_pattern [String,Regex] pattern to search for
+    # @return [String] paths to files found
+    #
     def dir_and_glob(glob_pattern)
```
Comment (roliveri) [inline lib/virtfs/find_class_methods.rb]: Here too, the following methods are for internal use.

**Review by roliveri — CHANGES_REQUESTED** [review]

**Comment by movitto** [comment]
@roliveri fixed the semantic issues and marked the noted methods as @api private meaning they can be hidden with the '--hide-api private' yardoc cli option

**Comment by movitto** [comment]
Not sure how this  PR was auto-closed and y I can't reopen...

File: lib/virtfs/context_switch_class_methods.rb [hunk 7]
```diff
@@ -67,6 +67,13 @@ def normalize_path(p, relative_to = nil)
       VfsRealFile.expand_path(p, relative_to || context.getwd).sub(/^[a-zA-Z]:/, "") # XXX
     end
 
+    # Invoke block withing the given filesystem context
+    #
+    # @param fs [VirtFS::FS] filesystem intstance through which to invoke block
+    # @param path [String] path to specify to block
+    #
+    # @raise [VirtFS::NotImplementedError] if filesystem method does not exist
+    #
```
Comment (roliveri) [inline lib/virtfs/context_switch_class_methods.rb:77]: I think this should be `@api private` as well.

File: lib/virtfs/context_switch_class_methods.rb [hunk 8]
```diff
@@ -75,6 +82,10 @@ def fs_call(fs, path = nil, &block)
       raise VirtFS::NotImplementedError.new(fs, err.name)
     end
 
+  # Invoke block using fully resolved filesystem path
+  #
+  # @see .fs_call
+  # @see Context#path_lookup
```
Comment (roliveri) [inline lib/virtfs/context_switch_class_methods.rb]: I think this should be `@api private` as well.

File: lib/virtfs/find_class_methods.rb [hunk 9]
```diff
@@ -1,11 +1,17 @@
...
+  # VirtFS Find Class representation - implements the core Ruby Find methods,
+  # dispatching to underlying mounted VirtFS filesystems
   module FindClassMethods
     #
     # Modified version of Find.find:
     # - Accepts only a single path.
     # - Can be restricted by depth - optimization for glob searches.
     # - Will work with VirtFS, even when it's not active.
     #
+    # @param path [String] starting director of the find
```
Comment (roliveri) [inline lib/virtfs/find_class_methods.rb]: typo: director should be directory

**Review by roliveri — CHANGES_REQUESTED** [review]
@movitto please don't squash commits, it makes it harder to review incremental changes.

Also, when documenting methods that are just re-implementations of their standard Ruby counterparts, I wonder if we should just refer to the standard Ruby documentation.

File: lib/virtfs/find_class_methods.rb (hunk 9 above)
Comment (movitto) [reply lib/virtfs/find_class_methods.rb]: @roliveri updated

File: lib/virtfs/context_switch_class_methods.rb (hunk 8 above)
Comment (movitto) [reply lib/virtfs/context_switch_class_methods.rb]: @roliveri updated

File: lib/virtfs/context_switch_class_methods.rb (hunk 7 above)
Comment (movitto) [reply lib/virtfs/context_switch_class_methods.rb:77]: @roliveri updated

**Comment by movitto** [comment]
@roliveri yes that makes sense (I believe we can link directly to it for the yardocs). Perhaps this can be done in a subsequent PR though so as not to hold things up.

**Comment by roliveri** [comment]
and yet, you still squashed the commits.

**Comment by movitto** [comment]
@roliveri srry thought you were referring to the other PR (virtfs-ext3), forgot if this one had separate commits to begin with (switching back and force between several virtfs projects, tricky to keep track!)

**Pull request merged**